job-runner/chain/traverser.go

```diff
@@ -81,209 +110,169 @@ func (f *traverserFactory) Make(jobChain proto.JobChain) (Traverser, error) {
-		t.chain.SetState(finalState)
-		t.chainRepo.Set(t.chain)
-
-		// Mark the request as finished in the Request Manager.
-		if err := t.rmc.FinishRequest(t.chain.RequestId(), finalState); err != nil {
-			t.logger.Errorf("problem reporting status of the finished chain: %s", err)
-		} else {
-			t.chainRepo.Remove(t.chain.RequestId())
-		}
-	}()
-
-	firstJob, err := t.chain.FirstJob()
-	if err != nil {
-		finalState = proto.STATE_FAIL
-		return err
+	// Find all the jobs we can start running. For a new job chain (not suspended),
+	// this'll end up being just the first job in the chain.
+	jobsToRun := t.chain.RunnableJobs()
+	if len(jobsToRun) == 0 {
+		// No runnable jobs means the chain must already be done.
+		return
 	}
 
-	// Set the starting state of the chain.
-	t.chain.SetState(proto.STATE_RUNNING)
-	t.chainRepo.Set(t.chain)
-
 	// Start a goroutine to run jobs. This consumes from the runJobChan. When
 	// jobs are done, they will be sent to the doneJobChan, which gets consumed
 	// from right below this.
 	go t.runJobs()
 
-	// Add the first job in the chain to the runJobChan.
-	t.logger.Infof("sending the first job (%s) to runJobChan", firstJob.Id)
-	t.runJobChan <- firstJob
-
-	// When a job finishes, update the state of the chain and figure out what
-	// to do next (check to see if the entire chain is done running, and
-	// enqueue the next jobs if there are any).
-JOB_REAPER:
-	for doneJob := range t.doneJobChan {
-		jLogger := t.logger.WithFields(log.Fields{"job_id": doneJob.Id, "sequence_id": doneJob.SequenceId, "sequence_retry_count": t.chain.SequenceRetryCount(doneJob)})
-
-		// Set the final state of the job in the chain.
-		t.chain.SetJobState(doneJob.Id, doneJob.State)
-		t.chainRepo.Set(t.chain)
-
-		// Check to see if the entire chain is done. If it is, break out of
-		// the loop on doneJobChan because there is no more work for us to do.
-		//
-		// A chain is done if no more jobs in it can run. A chain is
-		// complete if every job in it completed successfully.
-		done, complete := t.chain.IsDone()
-		if done {
-			close(t.runJobChan)
-			if complete {
-				t.logger.Infof("chain is done, all jobs finished successfully")
-				finalState = proto.STATE_COMPLETE
-			} else {
-				t.logger.Warn("chain is done, some jobs failed")
-				finalState = proto.STATE_FAIL
-			}
-			break
-		}
-
-		// done jobs are either complete or failed
-		if doneJob.State != proto.STATE_COMPLETE {
-			if t.chain.CanRetrySequence(doneJob) {
-				jLogger.Info("job did not complete successfully. retrying sequence.")
-				t.retrySequence(doneJob) // re-enqueue first job of failed sequence
-				continue JOB_REAPER
-			}
-
-			// If the failed job is not part of a retryable sequence, then ignore
-			// subsequent jobs. For example, if job B of A -> B -> C  fails, then C is
-			// not ran.
-			jLogger.Warn("job did not complete successfully")
-			continue JOB_REAPER
+	// Suspend running job chain if Job Runner shuts down.
```

Comment:
Can move this block to end of func, replacing the single `<-t.doneChan` line with this block sans the goroutine. This block is the complete "when does this func stop" code.